pkg/controller/jaeger/jaeger_controller.go

```diff
@@ -352,7 +352,12 @@ func (r *ReconcileJaeger) apply(ctx context.Context, jaeger v1.Jaeger, str strat
 	if err := r.applyHorizontalPodAutoscalers(ctx, jaeger, str.HorizontalPodAutoscalers()); err != nil {
 		// we don't want to fail the whole reconciliation when this fails
 		jaeger.Logger().WithError(tracing.HandleError(err, span)).Warn("failed to reconcile pod autoscalers")
-		return jaeger, nil
+	}
+
+	if jaeger.Spec.ServiceMonitor.Enabled != nil && *jaeger.Spec.ServiceMonitor.Enabled {
```

Comment:
I think we might need a safe-guard here. When the service monitor isn't available but the CR sets it as "enabled=true", we see this:

```
WARN[0060] failed to reconcile servicemonitor            error="no matches for kind \"ServiceMonitor\" in version \"monitoring.coreos.com/v1\"" instance=simple-prod namespace=default
```

We do detect already whether the cluster has support for this API:

```
$ make run
...
INFO[0001] Install prometheus-operator in your cluster to create ServiceMonitor objects  error="no ServiceMonitor registered with the API"
```

This is probably done during the bootstrap, so, we could just add to viper whether we should attempt to create service monitors if they are enabled.

Once we have that, it might make sense to set the default value of the "Enabled" flag to `true` in case the Prometheus Operator is available, and to `false` otherwise. This way, the flag is only needed in exceptional cases.

What do you think?

Reply:
Sounds good. I'm not sure if we should enable the ServiceMonitor by default, since it is changing default behavior, but i would leave this decision up to you.

Reply:
@kevinearls, @rubenvp8510, @objectiser, what do you think? One one hand, having two defaults might be confusing. On the other hand, it kinda makes sense to create a service monitor by default if the API for it exists. I think we have a similar behavior with the OAuth Proxy.

Reply:
@jpkrohling  if by two defaults you mean "If the prometheus operator is installed we create a service monitor by default, otherwise we don't" I think that would be fine, especially if we already have this behavior elsewhere.

Reply:
@jpkrohling I think maybe for 1.x we should keep the same behaviour, but possibly revisit the defaults for 2.x?

Reply:
Sorry misunderstood - yes I think it is ok to create if API exists, as long as we document that as the behaviour

Reply:
@sschne, looks like we have a decision. Do you know what's needed to implement this?

Reply:
@jpkrohling i see there has been done similar things with Elasticsearch Operator. So i detect whether the API is available via the autodetect package and add it to viper.

Reply:
I'd still keep the CR flag in place: people should be able to individually turn it off on an individual basis, perhaps because the instance is a dev/staging one that doesn't need to be monitored.

Reply:
The CLI flag is now in place with an default auto-detection of the prometheus-operator CRDs.

If the CR field is set to "enable" but the CRDs are not available or the global flag is explicitly disabled, it will issue a warning. This is handled in the jaeger-controller.

The admin services however, which are now dedicated Service objects, will always be created if the CR field is set to "enable", despite the status of the CLI flag.